pkg/fleet-manager/application_controller.go

```diff
@@ -0,0 +1,384 @@
+		return fmt.Errorf("failed to add a Watch for GitRepository: %v", err)
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &sourcev1a2.HelmRepository{}},
+		handler.EnqueueRequestsFromMapFunc(a.objectToApplicationFunc),
+	); err != nil {
+		return fmt.Errorf("failed to add a Watch for HelmRepository: %v", err)
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &kustomizev1.Kustomization{}},
+		handler.EnqueueRequestsFromMapFunc(a.objectToApplicationFunc),
+	); err != nil {
+		return fmt.Errorf("failed to add a Watch for Kustomization: %v", err)
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &helmv2b1.HelmRelease{}},
+		handler.EnqueueRequestsFromMapFunc(a.objectToApplicationFunc),
+	); err != nil {
+		return fmt.Errorf("failed to add a Watch for HelmRelease: %v", err)
+	}
+
+	return nil
+}
+
+func (a *ApplicationManager) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+	app := &applicationapi.Application{}
+	if err := a.Get(ctx, req.NamespacedName, app); err != nil {
+		if apiserrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, errors.Wrapf(err, "failed to get application %s", req.NamespacedName)
+	}
+
+	log := ctrl.LoggerFrom(ctx)
+	log = log.WithValues("application", klog.KObj(app))
+
+	patchHelper, err := patch.NewHelper(app, a.Client)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to init patch helper for application %s", req.NamespacedName)
+	}
+
+	defer func() {
+		patchOpts := []patch.Option{}
+		if err := patchHelper.Patch(ctx, app, patchOpts...); err != nil {
+			reterr = utilerrors.NewAggregate([]error{reterr, errors.Wrapf(err, "failed to patch application %s", req.NamespacedName)})
+		}
+	}()
+
+	// Add finalizer if not exist to void the race condition.
+	if !controllerutil.ContainsFinalizer(app, ApplicationFinalizer) {
+		controllerutil.AddFinalizer(app, ApplicationFinalizer)
+		return ctrl.Result{}, nil
+	}
+
+	// Define object key for fleet based on the current policy's destination
+	fleetKey := client.ObjectKey{
+		Namespace: app.Namespace,
+		// there only one fleet
+		Name: app.Spec.SyncPolicy[0].Destination.Fleet,
+	}
+	fleet := &fleetapi.Fleet{}
+	// Retrieve fleet object based on the defined fleet key
+	if err := a.Client.Get(ctx, fleetKey, fleet); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("fleet does not exist", "fleet", fleetKey)
+			return ctrl.Result{RequeueAfter: RequeueAfter}, nil
+		}
+
+		// Log error and requeue request if error occurred during fleet retrieval
+		log.Error(err, "failed to find fleet", "fleet", fleetKey)
+		return ctrl.Result{}, err
+	}
+
+	// label the fleet
+	if fleet.GetLabels() == nil {
+		fleet.SetLabels(make(map[string]string))
+	}
+	// todo: there is a bug here if more than one application refer the same fleet: the new one will override the old
+	//  but it is needed to reconcile application when the fleet' cluster is all joined
+	if fleet.GetLabels()[ApplicationLabel] != app.Name {
```

Comment:
Why do we need to set app labels to a fleet, in reality there could be multiple applications for a single fleet